### DISCUSSION TITLE
v4.0.39-alpha2

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#87] [Announcements]
The proposed v4.0.39 carries some significant additions, hence I want to introduce it as an alpha release.

I must stress that running this alpha is not going to obliterate history, configuration, nothing. It is (mostly) all adds. Deciding to run it, then switching back to @BJReplay is simple, and equally will not obliterate anything.

So what's the payload?

* Firstly getting rid of the scipy library, and good riddance. It was turning into dependency Hell after just a couple of days, so it's gone, and replaced with old-school math coding.
* Secondly more configuration options, specifically relating to attributes.

In the past, some have commented that having the detailed forecast breakdown as an attribute wasn't something they used. Fair enough. You can turn it off now.

All those forecast90s? They're fanciful. I don't want them! So turn them off... And why are the forecast50s replicated as attributes as well? Yeah, you get the idea. Turn 'em off.

A site breakdown? Seriously? Because @gcoan asked for it you're filling my storage? So turn it off...

<img width="418" alt="image" src="https://github.com/user-attachments/assets/494eb89a-4b94-47bd-8429-8b476b32d5e0">

<img width="413" alt="image" src="https://github.com/user-attachments/assets/461d133f-ae20-4ec7-a53e-3952d6a7b819">

I have tried to add translations for the languages that are present in the integration. German, French, that weirdo English, Slovak, Polish and Urdu. (Honestly, my Urdu is not that flash.) This was a Google Translate job, so feedback is most very welcome. The German and French made sense to me in my head, but elsewhere...

Because this is an alpha, I have hosted it as a repo on @autoSteve to _completely_ separate it from the BJReplay stream (we deffo don't want to cross the streams...) To implement it requires changing the custom HACS repository.

This is straightforward. Don't panic.

**_Do not remove your integration._**

Go to custom repositories in HACS.

<img width="266" alt="image" src="https://github.com/user-attachments/assets/72fe1c0a-67d0-4ea2-8955-3ac57bfb830e">

Remove BJReplay/ha-solcast-solar, and add autoSteve/dev-solcast-solar as an 'Integration' from https://github.com/autoSteve/dev-solcast-solar/.

Close that, and then find Solcast, then choose to download it. It won't offer a choice of version. You will be downloading v4.0.39-alpha2.

Restart HA.

(And un-cross your fingers. I know they're likely crossed... For the faint hearted, do take a backup first. I have hourly backups, not being faint hearted, just sensible.)

I have done this on my own (non-dev) install to prove that it works, so I am running my production HA from this very repo. I lost nothing, and gained much.

Reverting to BJReplay is the reverse. Change the custom repo as an 'Integration', download the integration, restart. And that will be precisely the procedure once this alpha is concluded. (With the exception being that reversion will download 'latest' 4.0.31, and not the most current beta if you were running it, so you'll need to re-download and select the desired version.)

**Comment by autoSteve** [comment]
Lols. I've broken it already. Working on it...

**Comment by autoSteve** [comment]
.ZIP replaced in the release with something that won't reset the options every restart. Sorry 'bout that. 🤦🏻

**Comment by gcoan** [comment]
installed the Alpha no problem and it seemed to upgrade my v8 config to the new v8 with no issues

Went and turned off the pesky estimate90 and hourly data I don't want, entities updated successfully.

Checked the log and no errors reported, but I did see that for both the initial load (when I was genuinely upgrading from the v8 pre-release to the new v8) and for the subsequent integration reload after I changed the config, the logfile contains the pesky misleading messages again:

```
2024-07-16 12:41:16.193 DEBUG (MainThread) [custom_components.solcast_solar] Options version 8
2024-07-16 12:41:16.193 DEBUG (MainThread) [custom_components.solcast_solar] Upgraded to options version 8
```

**Reply by autoSteve** [reply]
You are observant, Geoff, aren't you? And it made your eye twitch with PTSD didn't it?

But ironically, mate, this is all for you. Because you updated to a prior v8, which contained some new options, I had to force it to keep re-updating (except doing nothing on subsequent restarts). In the beta that will be gone. Your OCD will be fixed.

**Reply by gcoan** [reply]
> You are observant, Geoff, aren't you? And it made your eye twitch with PTSD didn't it?

Got to make it look like I'm adding some value to the party

> In the beta that will be gone. Your OCD will be fixed

Phew. It will rest better tonight 😀

**Comment by autoSteve** [comment]
For the actual 'changelog', see the branch https://github.com/BJReplay/ha-solcast-solar/tree/Daily-forecasts-site-breakdown. This repo is a mirror of that branch, just without any commit log (the "separation of streams"...) Feel free to shoehorn that into your install instead, should you know how if it makes you more comfortable with the code. They are _almost_ the same code (with two minor exceptions), and that branch will become the reviewed pull request to BJReplay, not this.

Minor exception one? That is currently giving @gcoan PTSD about the constant config version upgrade notifications on each integration start. It is the difference between a "7" and an "8" in the code... And ~~he~~ I really caused it, but it was for his own good.

And two? The BJReplay branch is currently ahead by a commit on two capitalisation fixes on the word "Watts".

**Comment by autoSteve** [comment]
The only "issue" I have noticed so far is that the forecast remaining kWh for the day can show as "-0.0", and this will be caused by the new spline math. I have corrected this, and will push a fix to the BJReplay branch, and not this alpha because so minor.

**Comment by autoSteve** [comment]
Does four downloads, plus me a quorum make? I'd honestly like more, but may not get more.

The silence for this alpha is eerie. I can't fault it... Can anyone? Then again, I can't read Pakistani/Persian Urdu to know if I stuffed that up.

The only v4.0.39 "fail" so far is for an unrelated and mostly usual not-read-the-readme from a new user. That and a rounding error for zero, which is fixed.

My year 12 cubic poly math seems to be working exactly like scipy did (good riddance), and the new options function for me like a bought one...

**Comment by autoSteve** [comment]
Alpha2 released. A minor oopsie on my part on startup fixed... Which caused it not to start. 🤦🏻

**Comment by autoSteve** [comment]
@ladaan, @Godzillahu, looping you in to this discussion.

**Comment by autoSteve** [comment]
Looping in @jeffb1974 to the alpha discussion.

**Reply by jeffb1974** [reply]
Alpha1 gave me the startup error, Alpha2 working well (so far) @autoSteve .

**Reply by autoSteve** [reply]
Sweet as. Play with the new attribute options to taste and report back!

**Reply by autoSteve** [reply]
_This_ is why I like alpha releases, by the way... A small cohort will brave the unexpected, and thank you!

**Comment by gcoan** [comment]
Alpha1 worked for me, alpha2 worked for me, don't know why I bothered upgrading, but I'm nice like that

**Comment by autoSteve** [comment]
Looping in @rholligan to the alpha discussion.

**Comment by rholligan** [comment]
Installed Alpha2 on my Home Assistant Green. Forecasts updating as expected, and configuration options available. No scipy issues :-).

Thanks Steve for your efforts and quick turnaround on this one!

**Reply by autoSteve** [reply]
Nice one.

It's killing me that you don't speak Urdu. I'm dying to know how good Google translate is! 🤣

**Comment by autoSteve** [comment]
Seven downloads, plus me of alpha2. Nice.

I've now switched to running the proposed v4.0.39 code, replacing the alpha. All fine. (And Geoff, the repeated config version upgrades are actually gone...)

**Comment by autoSteve** [comment]
FYI, v4.0.39 draft release is primed and ready, which will end the alpha, and switch back to the @BJReplay repo. More to come.

Unless anyone has grievances to report...

Given the whole CrowdStrike thing you'll forgive me for delaying for about ten hours while I sleep and then double check all work with fresh eyes... You Europeans will be tucked up in bed by that time, so it'll probably be there when you wake up much later...

Reverting to the BJReplay repo in HACS is straightforward, as noted in the original post, except that will download v4.0.31. Choose to re-download, and select v4.0.39 after I pull the trigger.

Lemme know if there are any issues doing so.

I'll advise here when v4.0.39 is up at BJReplay.

BTW, I will be "out-of-office" for a week come tomorrow. I'm heading to Tokyo to rubberneck at stuff, and hang out in odd bars, so responses may be delayed depending on how good the rubbernecking is. I'm tipping the rubbernecking will be epic. (Did someone say MARIO CARTS through the streets of Tokyo??? SEND IT!!!)

#84, #85, and more.

**Reply by BJReplay** [reply]
I will be preserving battery until late tonight European time (and then probably sleeping).

No new Github Out-of-office announcements or anything until I am back on 230V

**Reply by gcoan** [reply]
> until I am back on 230V

230V nominal

Side effect of all this solar generation and (so far today, 62kWh of export) is that I'm pushing the local grid voltage up significantly.  It's not been below 240V all day and has been running around 258V through midday.

Oh, sorry, its winter in Australia @autoSteve

**Reply by autoSteve** [reply]
Yeah. No 'free' power going begging tomorrow around here. DISMAL prediction. We appear to be caught in some kind of antarctic blast at the mo. Northern Hemisphere for me, baby. I'm migrating... Actually thinking about it, equitorial I could do. Except Cyclones. I've only done one cyclone and it cured my curiosity while trying to BBQ in the "lee" of a massive bloody branch dodging, whoa!, there's a sheet of steel whizzing by... You get the idea. So maybe not equitorial. I need to be migratory, which with the kids paid off and an empty nest I could do.

Sadly I can't bring the PowerWalls with... I'm solar powered as an individual, so a week away will be a consolation prize.

**Comment by isorin** [comment]
@autoSteve Enjoy Tokyo!
I'm sad we didn't keep SciPy, I did a bit of work with it and it was going well.
What was the problem, it wasn't available for some platforms?

Anyway, one change I want to do is to switch to "Trapezoidal rule" when doing the energy calculations (we are using "Midpoint rule" now).
https://en.wikipedia.org/wiki/Riemann_sum#Trapezoidal_rule

**Reply by autoSteve** [reply]
IKR @isorin? Frustrating.

Quite a few HA deployments using the integration could not download scipy. And not just the odd one or two...

Just code the math, as it sounds like a good idea... `spline.py` in the integration implements some numpy-like functions, like `diff()` that you'll need.

**Comment by autoSteve** [comment]
All, v4.0.39 is up.

To switch from @autoSteve, in HACS:

* Add the custom repo `https://github.com/BJReplay/ha-solcast-solar`
* Remove custom repo `https://github.com/autoSteve/dev-solcast-solar'
* Search for `solcast`, open it, download button at bottom right, which will get you back on v4.0.31
* Restart HA
* Back in hacs, re-download Solcast integration, turn on `Show beta versions`, select v4.0.39, download
* Restart HA
* Verify in logs that v4.0.39 is running

It may be possible to do this with one HA restart. I had some minor HACS weirdness, ending up on v4.0.38, so upgraded in the HA settings and restarted HA again.
